c/client.c.go

```diff
@@ -0,0 +1,223 @@
+	*codeOutC = C.CString(code)
+
+	go func() {
+		s := <-status
+		if s.Error != nil {
+			// TODO: stick error message somewhere conventional for C to read.
+			C.call_callback(ctxC, cb, nil, C.int(codes.ERR_SEND_TEXT_RESULT))
+		} else if s.OK {
+			C.call_callback(ctxC, cb, nil, C.int(codes.OK))
+		} else {
+			C.call_callback(ctxC, cb, nil, C.int(codes.ERR_UNKNOWN))
+		}
+	}()
+
+	return int(codes.OK)
+}
+
+//export ClientSendFile
+func ClientSendFile(ctxC unsafe.Pointer, clientPtr uintptr, fileName *C.char, length C.int, fileBytes unsafe.Pointer, codeOutC **C.char, cb C.callback) int {
+	client, err := getClient(clientPtr)
+	if err != nil {
+		return int(codes.ERR_NO_CLIENT)
+	}
+	ctx := context.Background()
+
+	// TODO: is there a way to avoid copying?
+	reader := bytes.NewReader(C.GoBytes(fileBytes, length))
+
+	code, status, err := client.SendFile(ctx, C.GoString(fileName), reader, false)
+	if err != nil {
+		return int(codes.ERR_SEND_TEXT)
+	}
+	*codeOutC = C.CString(code)
+
+	go func() {
+		s := <-status
+		if s.Error != nil {
+			// TODO: stick error message somewhere conventional for C to read.
+			C.call_callback(ctxC, cb, nil, C.int(codes.ERR_SEND_TEXT_RESULT))
+		} else if s.OK {
+			C.call_callback(ctxC, cb, nil, C.int(codes.OK))
+		} else {
+			C.call_callback(ctxC, cb, nil, C.int(codes.ERR_UNKNOWN))
+		}
+	}()
+
+	return int(codes.OK)
+}
+
+//export ClientRecvText
+func ClientRecvText(ctxC unsafe.Pointer, clientPtr uintptr, codeC *C.char, cb C.callback) int {
+	client, err := getClient(clientPtr)
+	if err != nil {
+		return int(codes.ERR_NO_CLIENT)
+	}
+	ctx := context.Background()
+
+	go func() {
+		msg, err := client.Receive(ctx, C.GoString(codeC), false)
+
+		if err != nil {
+			C.call_callback(ctxC, cb, nil, C.int(codes.ERR_RECV_TEXT))
+		}
+
+		data, err := ioutil.ReadAll(msg)
+		if err != nil {
+			C.call_callback(ctxC, cb, nil, C.int(codes.ERR_RECV_TEXT_DATA))
+		}
+		dataC := C.CBytes(data)
+		fileC := (*C.file_t)(C.malloc(C.sizeof_file_t))
+		*fileC = C.file_t{
+			length: C.int(len(data)),
+			data:   (*C.uint8_t)(dataC),
+		}
+
+		C.call_callback(ctxC, cb, unsafe.Pointer(fileC), C.int(codes.OK))
+	}()
+
+	return int(codes.OK)
+}
+
+//export ClientRecvFile
+func ClientRecvFile(ctxC unsafe.Pointer, clientPtr uintptr, codeC *C.char, cb C.callback) int {
+	client, err := getClient(clientPtr)
+	if err != nil {
+		return int(codes.ERR_NO_CLIENT)
+	}
+	ctx := context.Background()
+
+	go func() {
+		msg, err := client.Receive(ctx, C.GoString(codeC), false)
+
+		if err != nil {
+			C.call_callback(ctxC, cb, nil, C.int(codes.ERR_RECV_TEXT))
+		}
+
+		data, err := ioutil.ReadAll(msg)
+		if err != nil {
+			C.call_callback(ctxC, cb, nil, C.int(codes.ERR_RECV_TEXT_DATA))
+		}
+		dataC := C.CBytes(data)
+		fileC := (*C.file_t)(C.malloc(C.sizeof_file_t))
+		*fileC = C.file_t{
+			length: C.int(len(data)),
+			data:   (*C.uint8_t)(dataC),
+		}
+
+		C.call_callback(ctxC, cb, unsafe.Pointer(fileC), C.int(codes.OK))
```

Comment:
same comments as above in the `ClientRecvText()`. The memory is not much (`sizeof file_t`) which is just a few bytes - size of an int + size of a pointer. The `call_callback()` itself seem like a function that would return back here. So, can't we just call `C.free(fileC)` after this statement?

Reply:
This is a very good point. I think the issue is whether or not the caller should be partially responsible for freeing the data or fully responsible for it. Currently `fileC.data` should be freed by the caller as well as `fileC`. @bryanchriswhite is adding a single struct to group all arguments into one, and that could have a `release` function defined that frees both `fileC`, `fileC.data` as well as any other things passed in the arguments struct.